### PR TITLE
Generate QHub Costs via `infracost`

### LINKED ISSUE #1315: [ENH] - Add Infracost to QHub CI and CLI
### Feature description

Infracost is a tool that helps us retrieve the difference in monthly cloud cost estimates between the current and planned state of Terraform projects. With wide support for AWS, GCP & Azure, it can be integrated to our QHub CI to help us understand how our costs will change in case we make critical changes to our infrastructure. 

### Value and/or benefit

With the integration to QHub CLI, a user will be able to retrieve the total predicted costs for running QHub on a public cloud provider.

### Anything else?

- [Infracost](https://www.infracost.io/)

### PR DESCRIPTION
**Pull request by HarshCasper**
Fixes #1315

## Changes introduced in this PR:

- Adds cost report generation using Infracost via `qhub cost-report`

## Types of changes

What types of changes does your PR introduce?

_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

### Requires testing

- [X] Yes
- [ ] No

### In case you checked yes, did you write tests?

- [ ] Yes
- [X] No

## Documentation

Does your contribution include breaking changes or deprecations?
If so have you updated the documentation?

- [X] Yes, docstrings
- [ ] Yes, main documentation
- [ ] Yes, deprecation notices

## Further comments (optional)

Demo:

```sh
$ qhub cost --path=./stages/
       Cost Breakdown        
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━┓
┃               Name ┃ Cost ┃
┡━━━━━━━━━━━━━━━━━━━━╇━━━━━━┩
│ Total Monthly Cost │   73 │
│  Total Hourly Cost │  0.1 │
└────────────────────┴──────┘
           Resource Breakdown            
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃                         Name ┃ Number ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│         Total Detected Costs │     11 │
│    Total Supported Resources │      3 │
│ Total Un-Supported Resources │      1 │
│   Total Non-Priced Resources │      7 │
│ Total Usage-Priced Resources │      3 │
└──────────────────────────────┴────────┘
Access the dashboard here: https://dashboard.infracost.io/share/6ubpwjvv7y2vjch2stant14ifnzzzv3l
```

### CODE REVIEW

**Review by viniciusdc — COMMENTED** [review]
Looks good already for me, a quick note @HarshCasper do you think you can easily add a pytest section for this? if it's too much work, we can open this in a different issue.

Also, can you make `--path=...` an optional arg with default to `stages`?

**Review by viniciusdc — APPROVED** [review]
LGTM

File: qhub/cli/cost.py
Comment (iameskild) [inline qhub/cli/cost.py:8]: Could you add a little more info to the `help` command regarding what the difference `Resource Breakdown` categories means?

File: qhub/cost.py
Comment (iameskild) [inline qhub/cost.py:74]: `Cost ($)` perhaps?

Comment (iameskild) [reply qhub/cost.py:74]: Or can we change the currency value?

File: qhub/cost.py
Comment (iameskild) [inline qhub/cost.py:110]: @HarshCasper when I trying this out myself, I kept running into this issue:

```
Traceback (most recent call last):
  File "/Users/eskild/opt/miniconda3/envs/qhub-dev/bin/qhub", line 33, in <module>
    sys.exit(load_entry_point('qhub', 'console_scripts', 'qhub')())
  File "/Users/eskild/Documents/Quansight/qhub/qhub/__main__.py", line 7, in main
    cli(sys.argv[1:])
  File "/Users/eskild/Documents/Quansight/qhub/qhub/cli/__init__.py", line 55, in cli
    args.func(args)
  File "/Users/eskild/Documents/Quansight/qhub/qhub/cli/cost.py", line 15, in handle_cost_report
    infracost_report(args.path)
  File "/Users/eskild/Documents/Quansight/qhub/qhub/cost.py", line 110, in infracost_report
    print(f"Access the dashboard here: {data['shareUrl']}")
KeyError: 'shareUrl'
```

Would you mind having a look?

**Review by iameskild — CHANGES_REQUESTED** [review]
Hey @HarshCasper, I decided to try this tool out myself to get a feel for it and I ran into a few issues. This is still on the right track 👍 

I also opened an issue regarding how to setup and use this subcommand if you can get to it (or perhaps for the next sprint):
https://github.com/nebari-dev/nebari/issues/57

File: qhub/cost.py
Comment (iameskild) [reply qhub/cost.py:110]: FWIW, I was trying this locally, trying to get an estimate for `~/Documents/Quansight/qhub-deployment-quansight-beta`

Comment (HarshCasper) [reply qhub/cost.py:110]: Hi @iameskild 👋 

I tried running it and check what the error was. Fortunately it worked for me:

```sh
       Cost Breakdown        
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━┓
┃               Name ┃ Cost ┃
┡━━━━━━━━━━━━━━━━━━━━╇━━━━━━┩
│ Total Monthly Cost │   73 │
│  Total Hourly Cost │  0.1 │
└────────────────────┴──────┘
           Resource Breakdown            
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃                         Name ┃ Number ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│         Total Detected Costs │     11 │
│    Total Supported Resources │      3 │
│ Total Un-Supported Resources │      1 │
│   Total Non-Priced Resources │      7 │
│ Total Usage-Priced Resources │      3 │
└──────────────────────────────┴────────┘
Access the dashboard here: https://dashboard.infracost.io/share/cuy3fqeeb3kqvcufrme7c9hlekcnuxa7
```

Comment (HarshCasper) [reply qhub/cost.py:110]: Can you please run these two commands and see what the output is:

```sh
infracost --version
infracost configure get api_key
```

Comment (HarshCasper) [reply qhub/cost.py:110]: Hi @iameskild — I took a look at it and found the root cause. The JSON payload that we received did not have dashboards enabled in it, hence QHub errored out. 

To enable this, we have to enable the dashboards via:

```sh
infracost configure set enable_dashboard true
```

Can you run this and try it out once? Since we are giving the users control over their Infracost installation, I suggest we add it out in the docs to let the users know to enable dashboards. Or we can add it as a flag that the user needs to add in case they wish to use it. What do you think? cc: @viniciusdc

File: qhub/cost.py
Comment (HarshCasper) [reply qhub/cost.py:54]: Do you want a comment or an error message?

File: qhub/cost.py
Comment (HarshCasper) [reply qhub/cost.py:74]: @iameskild We can change currencies! However, do we want to prioritise the feature right now?

File: qhub/cost.py
Comment (kcpevey) [inline qhub/cost.py:116]: I think it would be neat to add the ability to write to csv and/or png, but that's just an idea. I don't want to hold up this PR for that.

File: qhub/cost.py
Comment (kcpevey) [inline qhub/cost.py:118]: shouldn't these print statements be logger.info statements?

File: qhub/cost.py
Comment (kcpevey) [reply qhub/cost.py]: I think just adding `(USD)` is a good solution

**Comment by kcpevey** [comment]
I'd also like to see more information on what each of these categories mean as part of the print output. Maybe a paragraph before or after the breakdown (or bullet points, whatever makes sense)

File: qhub/cost.py
Comment (iameskild) [reply qhub/cost.py]: Thanks for the update @HarshCasper! I tried it with `enable_dashboard true` and it worked. The dashboard link is very neat!

I think you bring up a good point about how much of the `infracost` setup process we should leave to the user. I'm inclined to have this tool handle the setup for the user. Handling the install with a `qhub cost-estimate` subcommand might be a too much to add right now but adding the flag to enable dashboards sounds like a reasonable addition.

And like you said, either way, we should document this :)

File: qhub/cost.py
Comment (iameskild) [reply qhub/cost.py]: I agree, adding a way to change currencies is not a priority but let's add a label. `USD` as Kim suggested sounds good to me :)

@HarshCasper, would you mind just creating an issue listing a few of the enhancement requests raised in this PR and any others that you might suggest? Thanks a lot!!

File: qhub/cost.py
Comment (iameskild) [reply qhub/cost.py:54]: Could we explain what went wrong so perhaps an error message.

File: qhub/cost.py
Comment (HarshCasper) [reply qhub/cost.py]: Added it 👍

File: qhub/cost.py
Comment (HarshCasper) [reply qhub/cost.py]: I guess its not a log statement but rather should be showcased on the CLI interface. Hence it needs to be a print statement in my opinion.

**Comment by HarshCasper** [comment]
> We need a documentation page describing how this works. Can you add that to the new repo?

It is being tracked here: https://github.com/nebari-dev/nebari/issues/57

**Comment by HarshCasper** [comment]
> I'd also like to see more information on what each of these categories mean as part of the print output. Maybe a paragraph before or after the breakdown (or bullet points, whatever makes sense)

Would it make sense to have it on docs rather than making the CLI interface cluttered?

File: qhub/cost.py
Comment (iameskild) [reply qhub/cost.py:130]: Perhaps `console.print` like you have above?

File: qhub/cli/cost.py
Comment (iameskild) [reply qhub/cli/cost.py:8]: Any updates?

**Review by iameskild — CHANGES_REQUESTED** [review]
Hey @HarshCasper, I think there are a few items that @kcpevey and I raised that might need to be addressed before merging. 

One of those items is reformatting or including explanations for the output table:
```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃                         Name ┃ Number ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│         Total Detected Costs │     11 │
│    Total Supported Resources │      3 │
│ Total Un-Supported Resources │      1 │
│   Total Non-Priced Resources │      7 │
│ Total Usage-Priced Resources │      3 │
└──────────────────────────────┴────────┘
```
Should these numbers add up? Are some of these related to the others? If so, how?

Also given that we are relying on node-pools (which fall under a `Usage Resource` and don't explicitly get captured here), could you please make a note that the `general` node-pool will always have one node running will add quite an additional charge.

As an example, on AWS, using an `m5.2xlarge` instance type for the general node-pool (currently the default), it will cost `$0.384/hour * 24 hours * 30 days = $276.84` additional per month.

We can make further enhancements to this tool later but we should also capture those in an issue. 

Thanks for your patience :)

File: qhub/cost.py
Comment (iameskild) [reply qhub/cost.py]: Perhaps `console.print`?

File: qhub/cost.py
Comment (iameskild) [inline qhub/cost.py:131]: Perhaps `console.print`?

Comment (HarshCasper) [reply qhub/cost.py:131]: This looks much better now 👍

**Comment by HarshCasper** [comment]
Hi @iameskild @kcpevey 

This is how the updated cost report looks like:

```sh
% qhub cost-estimate --path=/Users/harshcasper/code/quansight/qhub-deployment-quansight
         Cost Breakdown          
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┓
┃               Name ┃ Cost ($) ┃
┡━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━┩
│ Total Monthly Cost │       73 │
│  Total Hourly Cost │      0.1 │
└────────────────────┴──────────┘
           Resource Breakdown            
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃                         Name ┃ Number ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│         Total Detected Costs │     11 │
│    Total Supported Resources │      4 │
│ Total Un-Supported Resources │      0 │
│   Total Non-Priced Resources │      7 │
│ Total Usage-Priced Resources │      4 │
└──────────────────────────────┴────────┘
Access the dashboard here: 
https://dashboard.infracost.io/share/5c9spn9dv4sdqttfjlzvxcts9rlu7ib1

QHub rely upon node-pools which is a usage resource but doesn't get captured in 
the above report. A general node-pool will always have one node running will add
quite an additional charge. Please check in with your cloud provider to see the 
associated costs with node pools.                                               

 • Total Monthly Cost: The total monthly cost of the deployment of supported    
   resources.                                                                   
 • Total Hourly Cost: The total hourly cost of the deployment of supported      
   resources.                                                                   
 • Total Detected Costs: The total number of resources detected by Infracost.   
 • Total Supported Resources: The total number of resources supported by        
   Infracost.                                                                   
 • Total Un-Supported Resources: The total number of resources unsupported by   
   Infracost.                                                                   
 • Total Non-Priced Resources: The total number of resources that are not       
   priced.                                                                      
 • Total Usage-Priced Resources: The total number of resources that are priced  
   based on usage.                                                              
```

File: qhub/cost.py
Comment (viniciusdc) [reply qhub/cost.py]: I agree with @iameskild, we should let the user do the config by themselves which restricts the maintenance of this in the future.

**Comment by HarshCasper** [comment]
Thanks, @iameskild @viniciusdc and @kcpevey for all the reviews!

**Pull request merged**
